components/script/script_thread.rs

```diff
@@ -326,11 +326,58 @@ impl OpaqueSender<CommonScriptMsg> for Sender<MainThreadScriptMsg> {
     }
 }
 
+/// The state that a document managed by this script thread could be in.
+#[must_root]
+enum DocumentState {
+    /// A document that we are currently keeping.
+    /// The boolean flag indicates whether the script thread is keeping the document alive.
+    /// When documents are active, the flag is true, but if they become inactive we
+    /// set it to false, at which point the document can be garbage collected.
+    Kept(bool, JS<Document>),
```

Comment:
It might be nice to use a `enum DocumentReference { Strong, Weak }` instead of a boolean.